### DISCUSSION TITLE
Removing sub-shapes while in a sensor results in strange contact reports

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#433] [Q&A] [answered]
I'm getting some more interesting output from `OnContactAdded` and `OnContactRemoved` when dealing with compound shapes.

I'll preface by saying that I don't really understand how `SubShapeID` works, and no amount of staring at it seems to fix that, so it could just be that I'm fundamentally misusing them, through `SubShapeIDPair`, but I'm curious to hear your thoughts on it.

To illustrate my problem, I've added a new test to `Samples` (fork [here](https://github.com/mihe/JoltPhysics/blob/5301223a73387074d26c001e1941ad7bbd8f1517/Samples/Tests/General/SensorSubShapeTest.cpp)) that much like the one in #424 has a compound shape of three boxes pass through a static sensor, named `Left`, `Middle` and `Right`.

This time I'm also removing one of the shapes every time the body reaches the center of the sensor. I do this by rebuilding the entire `StaticCompoundShape` from scratch, which I'm sure isn't ideal, but that is none the less what I'm doing.

Here's what that looks like:

https://user-images.githubusercontent.com/4884246/217529623-55b50694-87c0-4915-b962-5e9c69aa8356.mp4

Here's what a heavily cleaned up version of the output looks like from the first two passes through the sensor:

```
Starting...

Body entering sensor
  OnContactAdded: [Sensor, Right]
  OnContactAdded: [Sensor, Middle]
  OnContactAdded: [Sensor, Left]

Removing the left shape
  OnContactAdded: [Sensor, Right]
  OnContactRemoved: [Sensor, Middle]
  OnContactRemoved: [Sensor, Left]

Body exiting sensor
  OnContactRemoved: [Sensor, Right]
  OnContactRemoved: [Sensor, Right]

Reached the far edge, turning back...

Body entering sensor
  OnContactAdded: [Sensor, Left]
  OnContactAdded: [Sensor, Middle]
  OnContactAdded: [Sensor, Right]

Removing the middle shape
  OnContactAdded: [Sensor, Right]
  OnContactRemoved: [Sensor, Middle]
  OnContactRemoved: [Sensor, Left]

Body exiting sensor
  OnContactRemoved: [Sensor, Right]
  OnContactRemoved: [Sensor, Right]

Reached the starting point.
```

As you might be able to see, the removal of the `Left` shape triggers some peculiar contact reports, where a new `Right` shape gets added and the old `Left` and `Middle` are removed. This results in the remaining overlapping shapes being the old `Right` and new `Right`. I assume this is because the new sub-shapes end up with IDs identical to the old ones, but shifted?

I guess the ideal output would be that it only reported the removed shape as having been removed, and then correctly report the two remaining shapes when they exit the sensor. So the first pass through the sensor would then look something like this:

```
Starting...

Body entering sensor
  OnContactAdded: [Sensor, Right]
  OnContactAdded: [Sensor, Middle]
  OnContactAdded: [Sensor, Left]

Removing the left shape
  OnContactRemoved: [Sensor, Left]

Body exiting sensor
  OnContactRemoved: [Sensor, Right]
  OnContactRemoved: [Sensor, Middle]

Reached the far edge, turning back...
```

I assume something like that would require `SubShapeID` to be stable across the two compound shapes though, and obviously there's no connection between the old and new shapes besides having identical user data, so I would frankly be happy if it simply reported all shapes as having been removed, only to immediately re-add the remaining shapes. So the first pass through the sensor would then look something like this:

```
Starting...

Body entering sensor
  OnContactAdded: [Sensor, Right]
  OnContactAdded: [Sensor, Middle]
  OnContactAdded: [Sensor, Left]

Removing the left shape
  OnContactRemoved: [Sensor, Right]
  OnContactRemoved: [Sensor, Middle]
  OnContactRemoved: [Sensor, Left]
  OnContactAdded: [Sensor, Middle]
  OnContactAdded: [Sensor, Right]

Body exiting sensor
  OnContactRemoved: [Sensor, Right]
  OnContactRemoved: [Sensor, Middle]

Reached the far edge, turning back...
```

**Comment by jrouwe** [comment]
I'm going to have to investigate this a bit more later (don't have much time right now).

Sub shape IDs are basically the concatenated bits from parent to child shape. So if you have CompoundShapeA containing CompoundShapeB containing MeshShapeC then the sub shape ID of a triangle in the mesh may be CCCCCCCBBAAA. The number of sub shapes in a compound is fixed, so it knows how many bits it needs, so shape A knows that it needs to take 3 bits. B has less shapes so only need 2 bits and a mesh shape has needs 7 bits to encode its triangle index (number of bits depends on number of triangles in the mesh).

This means that if you change the shape and you reuse an old sub shape ID that the bits don't make sense anymore (maybe you only need 2 bits for A now). I never ran into this issue because we don't change shapes on the fly for dynamic objects (we create new bodies if this happens). I need to think about this a bit more as by the time the system sends the callbacks, the old shapes are long gone, so there's no way you can get a ContactRemoved callback that makes sense at that point.

**Comment by mihe** [comment]
> I'm going to have to investigate this a bit more later (don't have much time right now).

No worries! Thanks for taking the time to reply.

To give a lot more context, I mentioned briefly in #424 that Godot signals when individual shapes enter/exit a sensor. This is in fact the only thing that the physics implementation reports back to the engine. The physics implementation reports what shape index has entered/exited a particular sensor, and [the engine layer on top](https://github.com/godotengine/godot/blob/2572f6800aef09bd6ea96f3b1c7a999a962eecb7/scene/3d/area_3d.cpp#L220-L284) keeps track of this and emits things like the body enter/exit signal when appropriate.

This more or less requires that the physics implementation accurately reports these shape enter/exit events, including when you remove shapes or whatever. This has been the source of some frustration, as I ironically also run into problems where this shape index is not what it used to be by the time I receive the `OnContactRemoved` from Jolt and invoke the callback.

My fallback plan is to just hack it and report a reference count as the shape index, similar to the count in `SensorTest::BodyAndCount`, which Jolt can obviously provide already. This would break some functionality for users who would depend on an accurate shape index in these events, but it would at least satisfy the engine I think. Considering some of the other things I've been unable to support so far it wouldn't be the end of the world. ~~I guess this would technically let me avoid having to disable `mUseManifoldReduction` as well.~~

**Comment by jrouwe** [comment, marked as answer]
I have thought about this some more and I think it's possible to properly support sending shape indices to Godot.

**Option 1:**
The OnContactAdded/Persisted/Removed pass in the user data of the shapes that are colliding. This introduces additional overhead since the user datas need to be fetched and more importantly they will need to be stored in the contact cache which will bloat the data structures. The data needs to be stored because at the time of the contact callbacks, the old shape is gone. Not a big fan of this option.

**Option 2:**
Supporting it in the Godot plugin. In JoltCollisionObject3D you'd add a ```Ref<Shape> previous_shape```. Whenever the shape changes, and previous_shape == nullptr you'd fetch the old shape of the body and store it here (purpose is to keep the shape that was used during the last simulation step). At the end of every physics simulation step you'd set previous_shape = nullptr to release the previous shape and free the memory.

In the ContactListener you treat the callbacks slightly different than before:

* OnContactAdded - this can just stay as is, you can use the sub shape IDs to resolve the shape index
* OnContactRemoved - if previous_shape != nullptr you'd use previous_shape to convert the sub shape ID to a shape index
* OnContactPersisted - if previous_shape != nullptr you'd first need to get the user data using the sub shape id on the old shape and then do the same on the new shape. If they're different then the same sub shape ID maps to different shapes so internally you would need to trigger a contact remove of the old sub shape and then contact add of the new sub shape. 

This was quite difficult to describe, so I'm hoping that you understand the idea. What do you think?

**Reply by mihe** [reply]
I'm happy to report that option 2 worked beautifully. In fact, since I cache the user data in my intermediate structure I ended up not really needing to deal with the user data in `OnContactRemoved`, so I was able to get away with only relying on the triggering of add/remove happening as a result of `OnContactPersisted`, and I would argue it [turned out](https://github.com/godot-jolt/godot-jolt/blob/51386453f71463478ab8f48d9f3340437849ec4d/src/jolt_contact_listener.cpp#L259-L286) pretty clean.

Thanks again!

**Comment by mihe** [comment]
> they will need to be stored in the contact cache which will bloat the data structures

I was afraid something like that would be needed. Definitely not worth it for something like this.

> In JoltCollisionObject3D you'd add a `Ref<Shape> previous_shape`

That's interesting! I've messed around with trying to cache various things across ticks to try to solve this, but never keeping the whole shape around to be able to query both with the same sub shape ID.

My only concern, at the top of my head, is what would happen if I rebuild the compound shape multiple times inbetween two steps, which is a possibility with the way things are structured now. Would Jolt send me contact reports for the intermediate shapes, or would it only be concerned with the shape at the time of the step? I assume it's the latter.

> This was quite difficult to describe, so I'm hoping that you understand the idea.

Yes! Thank you once again for such a in-depth look into the problem. I don't take it for granted.

I'll go ahead and close this issue, since it's clear there's no trivial way to solve this on the Jolt side of things. Hopefully your proposed solution pans out. If not, I have the fallback ready and working already.

**Comment by jrouwe** [comment]
Moving this to a discussion as the ideas here may be of future interest to other people.

> Would Jolt send me contact reports for the intermediate shapes, or would it only be concerned with the shape at the time of the step? I assume it's the latter.

Jolt will only ever generate contacts during the ```PhysicsSystem::Update``` function, so as long as you keep track of the shape that was used during the last call to that function, you will be able to resolve the sub shape IDs.
